File: 2187-minimum-time-to-complete-trips/2187-minimum-time-to-complete-trips.java
class Solution {
    public long minimumTime(int[] time, int totalTrips) {
        long start =1;
        long end = (long)getMin(time) * totalTrips;

        while(start < end ){
            long mid = start + ( end - start )/2;
            if(canComplete(time,totalTrips,mid)){
                end = mid;

            }
            else {
                start = mid + 1;
                       
            }
        }
        return start;
    
    }
    public boolean canComplete(int[]time, int totalTrips,long givenTime){
        int trip= 0;
        for(int t:time){
            trip+= givenTime/t;
            if(trip >=totalTrips){
                return true;
                 
            }
        }
        return false;
    }
    private int getMin(int[]time){
        int min = time[0];
        for(int t:time){
            min = Math.min(min,t);
        }
        return min;

    }
}

File: 2187-minimum-time-to-complete-trips/2187-minimum-time-to-complete-trips.cpp
class Solution {
public:
    long long minimumTime(vector<int>& time, int totalTrips) {
        long long left = 1;
        long long right =(long long)*max_element(time.begin(),time.end())* totalTrips;

        while(left<right){
            long long mid =left+(right-left)/2;
            if(canComplete(time,totalTrips,mid)){
                right =mid;
            }
            else{
                left = mid +1;
            }
        }
        return left;

    }

    private:
    bool canComplete(vector<int>& time, int totalTrips, long long givenTime){
        long long trips =0;
        for(int t:time){
            trips+= givenTime/t;
            if(trips>= totalTrips){
                return true;
            }
        }
        return false;
    }
};